models/shopify__discounts.sql

```diff
@@ -0,0 +1,90 @@
+with discount as (
+
+    select *
+    from {{ var('shopify_discount_code') }}
+),
+
+price_rule as (
+
+    select *
+    from {{ var('shopify_price_rule') }}
+),
+
+orders_aggregated as (
+
+    select *
+    from {{ ref('int_shopify__discounts__order_aggregates')}}
+),
+
+abandoned_checkouts_aggregated as (
+
+    select *
+    from {{ ref('int_shopify__discounts__abandoned_checkouts')}}
+),
+
+discount_price_rule_joined as (
+
+    select
+        discount.*,
+        price_rule.target_selection,
+        price_rule.target_type,
+        price_rule.title,
+        price_rule.usage_limit,
+        price_rule.value,
+        price_rule.value_type,
+        price_rule.allocation_limit,
+        price_rule.allocation_method,
+        price_rule.is_once_per_customer,
+        price_rule.customer_selection,
+        -- the below are NULL if customer_selection = all
+        price_rule.prereq_min_quantity,
+        price_rule.prereq_max_shipping_price,
+        price_rule.prereq_min_subtotal,
+        price_rule.prereq_min_purchase_quantity_for_entitlement,
+        price_rule.prereq_buy_x_get_this,
+        price_rule.prereq_buy_this_get_y,
+        price_rule.starts_at,
+        price_rule.ends_at,
+        price_rule.created_at as price_rule_created_at,
+        price_rule.updated_at as price_rule_updated_at
+
+    from discount
+    left join price_rule
+        on discount.price_rule_id = price_rule.price_rule_id
+        and discount.source_relation = price_rule.source_relation
+),
+
+aggregates_joined as (
+
+    select 
+        discount_price_rule_joined.*,
+        coalesce(orders_aggregated.total_order_discount_amount, 0) as total_order_discount_amount,
+        coalesce(abandoned_checkouts_aggregated.total_abandoned_checkout_discount_amount, 0) as total_abandoned_checkout_discount_amount,
+        coalesce(orders_aggregated.total_order_line_items_price, 0) as total_order_line_items_price,
+        coalesce(abandoned_checkouts_aggregated.total_abandoned_checkout_line_items_price, 0) as total_abandoned_checkout_line_items_price,
+        coalesce(orders_aggregated.total_order_shipping_cost, 0) as total_order_shipping_cost,
+        coalesce(abandoned_checkouts_aggregated.total_abandoned_checkout_shipping_price, 0) as total_abandoned_checkout_shipping_price,
+        coalesce(orders_aggregated.total_order_refund_amount, 0) as total_order_refund_amount,
+        coalesce(orders_aggregated.count_distinct_customers, 0) as count_distinct_customers,
+        coalesce(orders_aggregated.count_distinct_customer_emails) as count_distinct_customer_emails
+
+    from discount_price_rule_joined
+    left join orders_aggregated
+        on discount_price_rule_joined.code = orders_aggregated.code
+        and discount_price_rule_joined.source_relation = orders_aggregated.source_relation
+        -- in case one CODE can apply to both shipping and line items, percentages and fixed_amounts
```

Comment:
this is essentially necessary because we can only create these aggregated metrics at the `code` grain. >99% of the time, codes have a 1:1 relationship with the discount code PK (`discount_code_id`). Shopify recommends matching discount `codes` to price rule `titles` to keep this 1:1 relationship in tact.

the remaining <1% comes from cases in which the same code was used for different price rules, aka _kinds_ of discounts. these price rules can be slightly different -- do they apply to shipping or line items? are they percentages or fixed_amounts? if they apply to line items, do they apply to _all_ of them, or just select ones? 

the CASE WHEN logic i've added here incorporates the first two questions into our joins, so we're not mixing up/duplicating metrics. however, the third question (aka the `target_selection`), can throw a wrench in this. this field is not included in the underlying data we are aggregating, so if we join the aggregates with a discount code that was applied both across the board and to specific line items, we end up with some duplicate metrics that cannot be rolled up straightforwardly 

in the sample data, i only saw ONE case where this happened, and it was from 2020, so perhaps it will not really be a problem. felt that i had to call this out tho @fivetran-joemarkiewicz

Reply:
cool i included two tests -- one warning on non unique codes, one error on non unique discount_code_ids (this _really_ shouldn't have dupes)